### PR TITLE
Simplify isInnerClass

### PR DESCRIPTION
**Pull request by TurkeyMan**
Competing implementation for: https://github.com/dlang/phobos/pull/6757

### CODE REVIEW

File: std/traits.d [hunk 1]
```diff
@@ -2525,11 +2525,17 @@ Returns:
...
     static if (is(typeof(T.outer)))
-        enum isInnerClass = __traits(isSame, typeof(T.outer), __traits(parent, T))
-                         && (staticIndexOf!(__traits(allMembers, T), "outer") == -1);
+    {
+        template hasOuterMember(T...)
+        {
+            static if (T.length == 0)
+                enum hasOuterMember = false;
+            else
+                enum hasOuterMember = T[0] == "outer" || hasOuterMember!(T[1 .. $]);
```
Comment (thewilsonator) [inline std/traits.d]: Linear recursion template.

Comment (TurkeyMan) [reply std/traits.d]: It replaces a linear recursion template in a very deep import stack, so it's better than what's there.
I gotta hit the sack. You can log-N-ify it if you like....

**Comment by thewilsonator** [comment]
```
std/traits.d(2537:113)[warn]: Line is longer than 120 characters
posix.mak:554: recipe for target 'dscanner' failed
```

**Comment by thewilsonator** [comment]
I'll do an _actual_ \_\_trait tomorrow

**Comment by edi33416** [comment]
@TurkeyMan please see the comments on the previous PR.

I can close the other one, and you can get this though, but could you please add the unittest I was talking about in the [comment](https://github.com/dlang/phobos/pull/6757#issuecomment-445772614)?

Thanks,
Edi

File: std/traits.d (hunk 1 above)
Comment (TurkeyMan) [reply std/traits.d]: I've been thinking on it, and I don't think it's possible to improve on linear recursion :/
D needs C++'s `...` expression. That would eliminate 90% of recursive templates.

File: std/traits.d [hunk 2]
```diff
@@ -2525,11 +2525,20 @@ Returns:
...
+    {
+        bool hasOuterMember(string[] members...)
+        {
+            foreach (m; members)
+            {
+                if (m == "outer")
+                    return true;
+            }
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply std/traits.d]: Tested it... it really is!
CTFE `foreach` is also quite a lot better than `static foreach`... so we'll go with that.

**Comment by TurkeyMan** [comment]
I don't understand how I could have affected that test fail?

**Comment by thewilsonator** [comment]
Its green now.

**Comment by TurkeyMan** [comment]
It is. so...?

**Comment by wilzbach** [comment]
Just out of interest: how is this a simplification? It's more code now...

**Comment by TurkeyMan** [comment]
Nar, it's hundreds of lines less code, and a whole lot less imports.

**Comment by wilzbach** [comment]
> Nar, it's hundreds of lines less code, 

+13/-4 (!)

**Comment by TurkeyMan** [comment]
You've ignored the import stack.

**Comment by wilzbach** [comment]
std.meta is already imported at the top of std.traits

**Comment by TurkeyMan** [comment]
Instantiations are expensive, not imports. In particular, linear recursive instantiations.

**Comment by TurkeyMan** [comment]
```
isInnerClass 
- staticIndexOf
  - genericIndexOf (***linear recursive expansion***)
    - OldAlias
    - isSame
      - expectType
      - expectBool
```
Sure, it's not `max`, but it's still a lot, and I don't wanna mirror all that in druntime.

**Comment by TurkeyMan** [comment]
For reference, this is used from `emplace`, and I think it's terrible that `emplace` (whose job is to call a constructor), should cause a linear recursive expansion for every _member_ of every object, _recursively_(!), that you ever allocate.

**Pull request merged**